src/components/Firestore/DocumentEditor/store.tsx

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import get from 'lodash.get';
+import last from 'lodash.last';
+import produce from 'immer';
+import { Action, createReducer } from 'typesafe-actions';
+
+import * as actions from './actions';
+import { isMap, isArray, getParentPath } from '../utils';
+import { FirestoreAny, FirestoreMap } from '../models';
+
+const reducer = createReducer<FirestoreMap, Action>({})
+  .handleAction(actions.reset, (state, { payload }) => {
+    return payload;
+  })
+  .handleAction(
+    actions.addField,
+    produce((draft, { payload }) => {
+      const parent = get(draft, getParentPath(payload.path)) || draft;
+      if (isMap(parent)) {
+        parent[last(payload.path) as string] = payload.value;
```

Comment:
I think `last<string>()` is more safe, because it will check the type of the input array, vs cast the result.